[Feature] Added choice of units for margin placement to decoration items.

duncan-r opened the pull request:
- Updated dialogs with a QgsUnitSelectionWidget with choice of
  Percentage, pixels, or millimetres.
- Updated copyright, scalebar and north arrow to use the new
  unit selection.

duncan-r committed:
Updated decoration items to use QgsUnitSelectionWidget.

Applies to north arrow, copyright and scalebar.

- Changed dialogs to use QgsUnitSelectionWidget.
- Added QgsSymbolV2::OutputUnit enum to decorationitem.
- Updated to use mm, pixel, or percentage units.

duncan-r commented:
@nyalldawson I've added the QgsUnitSelectionWidget to allow a choice of margin units as we briefly discussed previously. I thought I'd notify you directly as you were involved in the changes last time (I hope this is ok?).

I see there have been some changes to the setup of the QgsDecorationItem classes and have tried to keep my updates in line with these. Let me know if you think there's any problems with the way it's setup and I will try and rectify.

duncan-r commented:
I'm not sure what's gone wrong here. Is this something I've done? Probably, but still a bit clueless?

SebDieBln commented:
> I'm not sure what's gone wrong here.

Have a look at [this error message](
https://travis-ci.org/qgis/QGIS/jobs/101243698#L192-L194).

> Is this something I've done? Probably, but still a bit clueless?

No, it is not your fault. It is just a timeout while downloading something.

Please note the difference between
* an _errored_ build, meaning a requested action could not be taken, and
* a _failed_ build, meaning an action did not yield an acceptable result

See [this example](https://travis-ci.org/qgis/QGIS/jobs/101271195) of a build that _failed_ due to [indentation errors](https://travis-ci.org/qgis/QGIS/jobs/101271195#L1012).

_Errored_ builds have the chance to succeed when they are re-run. In your fork of the repository you can trigger the rerunning of a job yourself, directly in Travis. In the upstream repository, where the jobs for PRs are run, normal users can not re-run jobs.
For _errored_ jobs in PRs it is best to simply start a new job by rebasing.

_Failed_ jobs generally have no chance to succeed by re-running them. Of course there can be exceptions to this rule.

nyalldawson commented:
I've just manually triggered a rebuild of this branch. Let's see how it goes...

duncan-r commented:
Thanks @SebDieBln for the explanation. I couldn't believe that it wasn't something stupid I'd done. Seemed more likely :). 

And thanks @nyalldawson. It seems happier this time around.

nyalldawson commented on src/app/qgsdecorationitem.h:
```diff
@@ -37,6 +37,14 @@ class APP_EXPORT QgsDecorationItem: public QObject
     };
 
+    //! Placement units
+    enum MarginUnits
```
Instead of creating a new enum, it would be better to reuse QgsSymbolV2::OutputUnit and add the "Percentage" mode to that enum. In any case Percentage will eventually need to be added to that enum for both labeling and layer effects, so may as well do that now and avoid the duplicate code.

nyalldawson commented on src/app/qgsdecorationcopyrightdialog.cpp:
```diff
@@ -49,6 +49,12 @@ QgsDecorationCopyrightDialog::QgsDecorationCopyrightDialog( QgsDecorationCopyrig
   cboPlacement->setCurrentIndex( cboPlacement->findData( mDeco.placement() ) );
   spnHorizontal->setValue( mDeco.mMarginHorizontal );
   spnVertical->setValue( mDeco.mMarginVertical );
+  QStringList mySelectionUnitItems;
```
If the units are changed to QgsSymbolV2::OutputUnit then this can be simplified (and made safer) by using a method like:
https://github.com/qgis/QGIS/blob/master/src/gui/effects/qgspainteffectwidget.cpp#L257

Then setting the current unit is done like https://github.com/qgis/QGIS/blob/master/src/gui/effects/qgspainteffectwidget.cpp#L283 and retrieving like https://github.com/qgis/QGIS/blob/master/src/gui/effects/qgspainteffectwidget.cpp#L343
(ignore the "mapUnitScale" code - that's not relevant for this, as map units won't be a choice in the widget).

nyalldawson commented on src/app/qgsdecorationitem.cpp:
```diff
@@ -65,6 +66,8 @@ void QgsDecorationItem::projectRead()
   QgsDebugMsg( "Entered" );
   mEnabled = QgsProject::instance()->readBoolEntry( mNameConfig, "/Enabled", false );
   mPlacement = static_cast< Placement >( QgsProject::instance()->readNumEntry( mNameConfig, "/Placement", static_cast< int >( mPlacement ) ) );
+  mMarginUnits = static_cast< MarginUnits >( QgsProject::instance()->readNumEntry( mNameConfig, "/MarginUnits", static_cast< int >( mMarginUnits ) ) );
```
If mMarginUnit is changed to QgsSymbolV2::OutputUnit then this should become:

    mMarginUnits = QgsSymbolLayerV2Utils::decodeOutputUnit( QgsProject::instance()->readEntry( mNameConfig, "/MarginUnits", QgsSymbolLayerV2Utils::encodeOutputUnit( mMarginUnits )  ) );

nyalldawson commented on src/app/qgsdecorationitem.cpp:
```diff
@@ -72,6 +75,7 @@ void QgsDecorationItem::saveToProject()
   QgsDebugMsg( "Entered" );
   QgsProject::instance()->writeEntry( mNameConfig, "/Enabled", mEnabled );
   QgsProject::instance()->writeEntry( mNameConfig, "/Placement", static_cast< int >( mPlacement ) );
+  QgsProject::instance()->writeEntry( mNameConfig, "/MarginUnits", static_cast< int >( mMarginUnits ) );
```
As per line 69, this would become

    QgsProject::instance()->writeEntry( mNameConfig, "/MarginUnits", QgsSymbolLayerV2Utils::encodeOutputUnit( mMarginUnits ) );

nyalldawson commented:
@duncan-r I've made some notes above. Give me a ping if you need clarification/assistance with any of them!

duncan-r commented:
@nyalldawson Thanks for the comprehensive suggestions. I'll have a bit of a look at it all and let you know if I have any issues.

duncan-r committed:
Added "Percentage" option to QgsSymboV2::OutputUnit.

- Added percentage to OutputUnit enum.
- Updated qgssymbolv2utils to include percentage.
- Updated qgsunitselectionwidget to include percentage.
Changed selection widget in its own commit

Added Percentage to OutputUnit enum.

Added percentage to switches in utils

duncan-r commented:
@nyalldawson I think this is now refactored into the setup that you suggested and is ready to push. I just wondered if there was any preference on default setting for the units. Styles and labels appear to both use millimetres, should I maintain consistency with them?

The section of the render() function that deals with placement and margin is getting quite big and has a fair bit of duplicate code. I did look at pulling this out into the base class a while back, but at present the way that placements are dealt with in the legacy code is slightly different in each derived class. It would also require a lot of variables to be passed back and forth - or a struct or additional members adding perhaps - and I didn't think it was wise to reformat a large chunk of several classes in my first contribution (which seems prudent based on how much more I know about the API now compared to when I started - which still amounts to about 20 files :) ).

Do you have any thoughts on this? It probably won't get done before the feature freeze anyway I imagine.

SebDieBln commented:
@duncan-r 
> I think this is now refactored into the setup that you suggested and is ready to push.

Then please push, so everybody can have a look at your updates. It is no problem pushing again when you updated the default settings for example. Even [rebasing](https://git-scm.com/docs/git-rebase) works well with PRs. So don't be afraid to push many commits, they can all be bundled together later.

duncan-r committed:
Set default margin units to MM

duncan-r committed:
Merge branch 'master' into feature_decoritem_marginunit

nyalldawson commented:
I've manually merged this - thanks! 

FYI - I did a quick follow up commit (https://github.com/qgis/QGIS/commit/9dccce87ee65d92d5124b1e9a6e3b5926ca1e037) to help speed up merging this before freeze. To explain the changes I made:

- Fix dialog tab order: QgsUnitSelectionWidget has some oddness with tab orders. You need to set the FocusPolicy of the widget to "Strong" for it to work correctly.
- Avoid use of hardcoded enum int conversion: It's safer/better to use the enum values rather than ints in switches. It helps readability, is safer if enum items are reordered in future, and allows searching for all the places a particular enum value is used.
- Remove some unneeded implicit casts: there was a bunch of explicit int/float conversions which didn't need to be in the calculations, since they'd happen implicitly. This was just to aid readability.
- Doc tweaks: just reworded the doc string for Percentage, as it won't always refer to percent of canvas size.

Just minor tweaks really. Once again, great work!

duncan-r commented:
@nyalldawson Thanks for your guidance in getting this properly sorted out. I've had a look at your changes. Sorry some, like the hardcoded enums, were slightly stupid oversights!

I'm intrigued by the unneeded explicit casts, as I expected implicit behaviour to start with, but was getting some strange results that seemed to be fixed by putting them in?

Oh well. Thanks again for all your help.
